apps/heft/src/cli/actions/BuildAction.ts

```diff
@@ -108,4 +240,25 @@ export class BuildAction extends HeftActionBase<IBuildActionData, BuildHooks> {
       watchMode: this._watchFlag.value
     };
   }
+
+  protected async _runPhaseWithLogging(buildPhaseName: string, buildPhase: IBuildPhase): Promise<void> {
+    if (buildPhase.hooks.run.isUsed()) {
+      await this._runWithLogging(buildPhaseName, async () => await buildPhase.hooks.run.promise());
+    }
+  }
+
+  private async _runWithLogging(buildPhaseName: string, fn: () => Promise<void>): Promise<void> {
+    this.terminal.writeLine(` ---- ${buildPhaseName} started ---- `);
+    const startTime: number = performance.now();
+    let finishedLoggingWord: string = 'finished';
+    try {
+      await fn();
+    } catch (e) {
+      finishedLoggingWord = 'encountered an error';
+      throw e;
+    } finally {
+      const executionTime: number = Math.round(performance.now() - startTime);
+      this.terminal.writeLine(` ---- ${buildPhaseName} ${finishedLoggingWord} (${executionTime}ms) ---- `);
```

Comment:
This is fine for now. 

But for the future, execution times should be tracked broadly and summarized in a report at the end. The person who is optimizing perf needs to see all the numbers in one place. And they also need the ability to collect timings for arbitrary code spans, as they drill down into problem areas. It's not realistic to expect that build phases will be the relevant unit of granularity.

A while ago I was investigating API Extractor perf issues and started work on a `ProfileTimer` system along these lines.  The work in progress is in this branch: https://github.com/microsoft/rushstack/commit/c8b39d46e6e55152e6df247ed63489651edf3d49  Maybe we could reuse that for Heft.

Reply:
We already report the total time at the end, but we can certainly add more fidelity to that report in the future. That'd be pretty useful.